contracts/contracts/lib/LibSubnetActor.sol

```diff
@@ -151,7 +153,36 @@ library LibSubnetActor {
         emit SubnetBootstrapped(s.genesisValidators);
 
         // register adding the genesis circulating supply (if it exists)
-        IGateway(s.ipcGatewayAddr).register{value: s.genesisCircSupply}(s.genesisCircSupply);
+        registerInGateway(0);
+    }
+
+    function registerInGateway(uint256 collateral) internal {
+        SubnetActorStorage storage s = LibSubnetActorStorage.appStorage();
+
+        uint256 genesisCircSupply = s.genesisCircSupply;
+
+        bool supplySourceNative = s.supplySource.isNative();
+        bool collateralSourceNative = s.collateralSource.isNative();
+
+        // this method is unnecessarily handling different cases because subnet actor needs
+        // to "register" in gateway and different token types needs to be attached or approved.
+        // TODO: it's known that having gateway holding all subnets' funds is insecure, this
+        // TODO: can be removed once contract redesign is in place.
+        uint256 msgValue = 0;
+
+        if (!supplySourceNative) {
+            s.supplySource.increaseAllowance(s.ipcGatewayAddr, genesisCircSupply);
+        } else {
+            msgValue += genesisCircSupply;
+        }
+
+        if (!collateralSourceNative) {
+            s.collateralSource.increaseAllowance(s.ipcGatewayAddr, collateral);
+        } else {
+            msgValue += collateral;
+        }
```

Comment:
This is leaky as mentioned elsewhere. I'm going to take a stab to contain this complexity inside the AssetHelper.